lib/elixir/lib/calendar/duration.ex

```diff
@@ -197,4 +197,144 @@ defmodule Duration do
+      iex> Duration.from_iso8601("P3Y-2MT3H")
+      {:ok, %Duration{year: 3, month: -2, hour: 3}}
+      iex> Duration.from_iso8601("-P3Y2MT3H")
+      {:ok, %Duration{year: -3, month: -2, hour: -3}}
+      iex> Duration.from_iso8601("-P3Y-2MT3H")
+      {:ok, %Duration{year: -3, month: 2, hour: -3}}
+
+  """
+  @spec from_iso8601(String.t()) :: {:ok, t} | {:error, String.t()}
+  def from_iso8601("P" <> duration_string) do
+    parse(duration_string, %{}, "", false)
+  end
+
+  def from_iso8601("-P" <> duration_string) do
+    case parse(duration_string, %{}, "", false) do
+      {:ok, duration} ->
+        {:ok, negate(duration)}
+
+      error ->
+        error
+    end
+  end
+
+  def from_iso8601(_) do
+    {:error, "invalid duration string"}
+  end
+
+  @doc """
+  Same as parse/1 but raises an ArgumentError.
+
+  ## Examples
+
+      iex> Duration.from_iso8601!("P1Y2M3DT4H5M6S")
+      %Duration{year: 1, month: 2, day: 3, hour: 4, minute: 5, second: 6}
+
+  """
+  @spec from_iso8601!(String.t()) :: t
+  def from_iso8601!(duration_string) do
+    case from_iso8601(duration_string) do
+      {:ok, duration} ->
+        duration
+
+      {:error, reason} ->
+        raise ArgumentError, "failed to parse duration. reason: #{inspect(reason)}"
+    end
+  end
+
+  defp parse(<<>>, duration, "", _), do: {:ok, new!(Enum.into(duration, []))}
+
+  defp parse(<<c::utf8, rest::binary>>, duration, buffer, is_time)
+       when c in ?0..?9 or c in [?., ?-] do
+    parse(rest, duration, <<buffer::binary, c::utf8>>, is_time)
+  end
+
+  defp parse(<<"Y", rest::binary>>, duration, buffer, false) do
+    parse(:year, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"M", rest::binary>>, duration, buffer, false) do
+    parse(:month, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"W", rest::binary>>, duration, buffer, false) do
+    parse(:week, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"D", rest::binary>>, duration, buffer, false) do
+    parse(:day, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"T", _::binary>>, _duration, _, true) do
+    {:error, "time delimiter was already provided"}
+  end
+
+  defp parse(<<"T", rest::binary>>, duration, _buffer, false) do
+    parse(rest, duration, "", true)
+  end
+
+  defp parse(<<"H", rest::binary>>, duration, buffer, true) do
+    parse(:hour, rest, duration, buffer, true)
+  end
+
+  defp parse(<<"M", rest::binary>>, duration, buffer, true) do
+    parse(:minute, rest, duration, buffer, true)
+  end
+
+  defp parse(<<"S", rest::binary>>, duration, buffer, true) do
+    parse(:second, rest, duration, buffer, true)
+  end
+
+  defp parse(<<c::utf8, _::binary>>, _, _, _) do
+    {:error, "unexpected character: #{<<c>>}"}
+  end
+
+  defp parse(unit, _string, duration, _buffer, _is_time) when is_map_key(duration, unit) do
+    {:error, "#{unit} was already provided"}
+  end
+
+  defp parse(:second, string, duration, buffer, is_time) do
+    case Float.parse(buffer) do
```

Comment:
There is `Calendar.ISO.parse_microsecond`, which we should use here instead. In fact, my proposal is to add `Calendar.ISO.parse_duration` that returns a keyword list of the units, and in this module you call `{:ok, fields} = Calendar.ISO.parse_duration(...); struct!(Duration, fields)`.

Reply:
> my proposal is to add Calendar.ISO.parse_duration 

I am in favour of this, but it'll be awkward to support the negative duration prefix in Calendar.ISO, given it depends on `Duration.negate/1`. While avoiding a circular dependency between Calendar.ISO and Duration, I dont really see how we could make it so both `Calendar.ISO.parse_duration` and `Duration.from_iso8601` could support the negative `-P` prefix 🤔 of course we can just not support it, but it'd be an unfortunate reason for doing so :)

Reply:
@tfiedlerdejanze you could also implement it by traversing the keyword list and negating all components afterwards, but i agree it is best to skip it for now.